rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,395 @@
+| feature | simple-package-paths |
+| --- | --- |
+| start-date | 2022-09-02 |
+| author | Silvan Mosberger (@infinisil) |
+| co-authors | Robert Hensing (@roberth) |
+| pre-RFC reviewers | Thomas Bereknyei (@tomberek), John Ericson (@Ericson2314), Alex Ameen (@aakropotkin) |
+| shepherd-team | @phaer @06kellyjac @aakropotkin @piegamesde |
+| shepherd-leader | - |
+| related-issues | https://github.com/NixOS/nixpkgs/pull/237439, https://github.com/NixOS/nixpkgs/pull/211832 |
+
+# Summary
+[summary]: #summary
+
+Auto-generate trivial top-level attribute definitions in Nixpkgs' `pkgs/top-level/all-packages.nix` from a directory structure that matches the attribute name.
+This makes it much easier to contribute new packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `all-packages.nix`.
+
+# Motivation
+[motivation]: #motivation
+
+- It is not obvious to package contributors where to add files or which ones to edit. These are very common questions:
+  - Which directory should my package definition go in?
+  - What are all the categories and do they matter?
+  - What if the package has multiple matching categories?
+  - Why can't I build my package after adding the package file?
+  - Where in `all-packages.nix` should my package go?
+- Figuring out where an attribute is defined is a bit tricky:
+  - First one has to find the definition of it in `all-packages.nix` to see what file it refers to
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even locks up editors due to the file size
+  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
+- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC consists of two parts, each of which is implemented with a PR to Nixpkgs.
+These PR's should be done after a release to maximize the testing period and minimize merge conflicts.
+
+## PR 1: The directory structure
+
+This part establishes the new directory structure in Nixpkgs.
+This directory structure is internal to Nixpkgs and not exposed as public interface.
+This directory structure must be documented in the Nixpkgs manual.
+This PR will be backported to the stable release in order to ensure that backports of new packages work.
+
+### File structure
+
+Create the initially-empty _base directory_ `pkgs/by-name` in Nixpkgs.
```

Comment:
Nitpick 1: git won't let you do that, so move `hello` in.

Nitpick 2: we barely use the term _base directory_ in this document and we should not refer to it outside this document, because that would be ambiguous. We may as well use `pkgs/by-name` in the remaining places.

```suggestion
Create the initially-empty `pkgs/by-name` directory in Nixpkgs, and migrate the `hello` package into it.
```